src/plugins/content_management/server/event_stream/es/es_event_stream_client.ts

```diff
@@ -0,0 +1,202 @@
+        match_all: {},
+      },
+      sort,
+      size: limit,
+    });
+    const events = res.hits.hits.map((hit) => dtoToEvent(hit._source!));
+
+    return events;
+  }
+
+  public async filter(
+    options: EventStreamClientFilterOptions
+  ): Promise<EventStreamClientFilterResult> {
+    const esClient = await this.deps.esClient;
+    const topLevelNodes: KueryNode[] = [];
+
+    if (options.subject && options.subject.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(
+          options.subject.map(([type, id]) =>
+            !id
+              ? nodeBuilder.is('subjectType', type)
+              : nodeBuilder.and([
+                  nodeBuilder.is('subjectType', type),
+                  nodeBuilder.is('subjectId', id),
+                ])
+          )
+        )
+      );
+    }
+
+    if (options.object && options.object.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(
+          options.object.map(([type, id]) =>
+            !id
+              ? nodeBuilder.is('objectType', type)
+              : nodeBuilder.and([
+                  nodeBuilder.is('objectType', type),
+                  nodeBuilder.is('objectId', id),
+                ])
+          )
+        )
+      );
+    }
+
+    if (options.predicate && options.predicate.length) {
+      topLevelNodes.push(
+        nodeBuilder.or(
+          options.predicate.map(([type, attributes]) => {
+            if (!attributes) {
+              return nodeBuilder.is('predicate', type);
+            }
+
+            const attributeNodes = nodeBuilder.and(
+              Object.entries(attributes).map(([key, value]) => {
+                return nodeBuilder.is(`payload.${key}`, value);
+              })
+            );
+
+            if (!type) return attributeNodes;
+
+            return nodeBuilder.and([nodeBuilder.is('predicate', type), attributeNodes]);
+          })
+        )
+      );
+    }
+
+    if (options.from) {
+      const from = new Date(options.from).toISOString();
+      const node = nodeBuilder.range('@timestamp', 'gte', from);
+
+      topLevelNodes.push(node);
+    }
+
+    if (options.to) {
+      const to = new Date(options.to).toISOString();
+      const node = nodeBuilder.range('@timestamp', 'lte', to);
+
+      topLevelNodes.push(node);
+    }
+
+    const query = toElasticsearchQuery(nodeBuilder.and(topLevelNodes));
+    const size = options.limit ?? 100;
+    const request: estypes.SearchRequest = {
+      index: this.#names.dataStream,
+      query,
+      sort,
+      size,
+    };
+
+    if (options.cursor) {
+      request.search_after = JSON.parse(options.cursor);
+    }
+
+    const res = await esClient.search<EsEventStreamEventDto>(request);
+    const events = res.hits.hits.map((hit) => dtoToEvent(hit._source!));
+    const lastHit = res.hits.hits[res.hits.hits.length - 1];
+
+    let cursor: string = '';
+
+    if (events.length >= size && lastHit && lastHit.sort) {
+      cursor = JSON.stringify(lastHit.sort);
```

Comment:
Haven't done much pagination in ES but that's all that's needed to paginate?

It seems that we might want to return the point in time too? (https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html)

Reply:
> ... that's all that's needed to paginate?

Yup. (That is called [`search_after` pagination](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#search-after).)

> It seems that we might want to return the point in time too?

I didn't think that PIT is needed here. As that would search through *old* data, i.e. the data when PIT search is initiated. And would be slower, too. But maybe I am missing something. Happy to discuss further online or in the thread.